packages/flutter/lib/src/painting/gradient.dart

```diff
@@ -16,22 +17,44 @@ class _ColorsAndStops {
   final List<double> stops;
 }
 
-_ColorsAndStops _interpolateColorsAndStops(List<Color> aColors, List<double> aStops, List<Color> bColors, List<double> bStops, double t) {
-  assert(aColors.length == bColors.length, 'Cannot interpolate between two gradients with a different number of colors.'); // TODO(ianh): remove limitation
-  assert((aStops == null && aColors.length == 2) || (aStops != null && aStops.length == aColors.length));
-  assert((bStops == null && bColors.length == 2) || (bStops != null && bStops.length == bColors.length));
-  final List<Color> interpolatedColors = <Color>[];
-  for (int i = 0; i < aColors.length; i += 1)
-    interpolatedColors.add(Color.lerp(aColors[i], bColors[i], t));
-  List<double> interpolatedStops;
-  if (aStops != null || bStops != null) {
-    aStops ??= const <double>[0.0, 1.0];
-    bStops ??= const <double>[0.0, 1.0];
-    assert(aStops.length == bStops.length);
-    interpolatedStops = <double>[];
-    for (int i = 0; i < aStops.length; i += 1)
-      interpolatedStops.add(ui.lerpDouble(aStops[i], bStops[i], t).clamp(0.0, 1.0));
-  }
+List<double> _impliedStops(int number) {
+  assert(number >= 2, 'colors list must have at least two colors');
+  final double separation = 1.0 / (number - 1);
+  return List<double>.generate(
+    number,
+        (int index) => index * separation,
+    growable: false,
+  );
+}
+
+Color _sample(List<Color> colors, List<double> stops, double t) {
+  if (t <= stops.first)
+    return colors.first;
+  if (t >= stops.last)
+    return colors.last;
+  final int index = stops.lastIndexWhere((double s) => s <= t);
+  return Color.lerp(
+      colors[index], colors[index + 1],
+      (t - stops[index])/(stops[index + 1] - stops[index])
+  );
+}
+
+_ColorsAndStops _interpolateColorsAndStops(
+    List<Color> aColors, List<double> aStops, List<Color> bColors, List<double> bStops, double t) {
```

Comment:
nit: add some line breaks?

```suggestion
    List<Color> aColors, 
    List<double> aStops, 
    List<Color> bColors,
    List<double> bStops, 
    double t,
) {
```